module/state_synchronization/execution_data_requester.go

```diff
@@ -0,0 +1,111 @@
+package state_synchronization
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/module/util"
+	"github.com/onflow/flow-go/network"
+)
+
+// ExecutionDataRequester requests ExecutionData for given root CIDs from the network.
+type ExecutionDataRequester struct {
+	blockExchange  network.BlockExchange
+	activeRequests sync.Map
+	serializer     *serializer
+}
+
+// RequestExecutionData requests the execution data for the given root CID.
+func (s *ExecutionDataRequester) RequestExecutionData(ctx context.Context, rootCid cid.Cid) ExecutionDataRequest {
+	req, exists := s.activeRequests.LoadOrStore(rootCid, &executionDataRequestImpl{
+		cid:  rootCid,
+		done: make(chan struct{}),
+	})
+	request := req.(*executionDataRequestImpl)
+	if !exists {
+		go s.handleRequest(ctx, request)
+	}
+
+	return request
+}
+
+func (s *ExecutionDataRequester) getData(ctx context.Context, session network.BlockExchangeFetcher, cids []cid.Cid) (reader, error) {
+	pendingCids := atomic.NewUint32(uint32(len(cids)))
+	indices := make(map[cid.Cid]int)
+	for i, cid := range cids {
+		indices[cid] = i
+	}
+
+	buf := make([]byte, (len(cids)-1)*MAX_BLOCK_SIZE)
+	var lastBlock blocks.Block
+
+	done, err := session.GetBlocks(cids...).ForEach(func(b blocks.Block) {
+		i := indices[b.Cid()]
+		if i == len(cids)-1 {
+			lastBlock = b
+			return
+		}
```

Comment:
Why treating the last block so special? 

Why not just copying the raw bytes into `buf`?